Extend samples plugin functionality

donat opened the pull request:
This PR makes the follwowing changes
- Introduces `isPromoted()` property to filter samples from samples index
- Allows samples to define dsl-specific exemplar tests
- Allows samples to declare explicit sanity checks (useful for failing examples)

big-guy commented on subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/internal/SamplesDocumentationPlugin.java:
```diff
@@ -415,7 +429,9 @@ private void realizeSamples(SamplesInternal extension, ObjectFactory objects, Ta
                 exemplarBinary.getTestedWorkingDirectory().convention(extension.getTestedInstallRoot().dir(toKebabCase(sample.getName()) + "/" + dsl.getConventionalDirectory())).disallowChanges();
                 exemplarBinary.getTestsContent().from(objects.fileCollection().from((Callable<FileTree>) () -> project.zipTree(binary.getZipFile())).builtBy(binary.getZipFile())); // TODO: zipTree should be lazy
                 exemplarBinary.getTestsContent().from(sample.getSampleDirectory().dir("tests"));
+                exemplarBinary.getDsl().value(dsl);
```
WDYT as an alternative:
- Don't add DSL to the install task since we end up using it to filter out content we don't want
- Change the convention to look for conf files in tests/ (not recursively) and conf files in tests/$dsl
- Make sure we flatten these into a single directory when installing to the install directory for test

donat replied:
In general, I like a good idea. So, our proposal would end up looking like this:
```
mySample
  |- common
  |- java
  |- kotlin
  |- test
     |- java
     |- kotlin
     |- common.sample.conf
```
How about a flat folder structure? Ultimately a complex sample could look like this:
```
mySample
  |- common
  |- java
  |- kotlin
  |- test-common
  |- test-java
  |- test-kotlin
```

donat replied:
> Make sure we flatten these into a single directory when installing to the install directory for test
I don't think the Exemplar tests would otherwise.

donat replied:
I've implemented approach.

big-guy commented on subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/internal/SamplesDocumentationPlugin.java:
```diff
@@ -90,8 +98,12 @@ public void apply(Project project) {
         // TODO: This could be lazy if we had a way to make the TaskContainer require evaluation
-        FileCollection generatedTests = createGeneratedTests(tasks, objects, layout);
-        extension.getBinaries().withType(SampleExemplarBinary.class).all(binary -> binary.getTestsContent().from(generatedTests));
+        FileCollection generatedSanityCheckTest = createGeneratedTests(tasks, objects, layout);
```
Maybe turn this around a bit...

- Have two tasks that generate sanityCheck and "negative" sanityCheck.  We'd have two different FileCollections here.
- Rename `getExplicitSanityCheck` to something like `getExpectedToFail` (default false)
- And then do something like:
```
binary.getTestsContent().from(binary.getExpectedToFail().flatMap(expectedToFail -> expectedToFail ? generatedNegativeTests : generatedTests))
```

donat replied:
I know my initial idea was to create negative sanity checks, but then I thought maybe there's a better alternative. How about the client can explicitly declare what the sanity check does. In that case, all negative examples could simply add a `sanityCheck.sample.conf` file with a `expect-failure: true` configuration.

big-guy commented on subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/internal/tasks/GenerateSanityCheckNegativeTests.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.FileWriter;
+import java.io.IOException;
+
+public abstract class GenerateSanityCheckNegativeTests extends DefaultTask {
```
is this used anywhere yet?

donat replied:
No, it's a leftover as it seemed cleaner to use explicit sanity checks.

donat commented on subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/Sample.java:
```diff
@@ -5,6 +5,9 @@
 import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+
+import org.gradle.api.provider.Property;
```
Yes. Removed.

donat commented on subprojects/gradle-guides-plugin/src/main/java/org/gradle/docs/samples/internal/tasks/InstallSample.java:
```diff
@@ -26,11 +33,32 @@ protected FileTree getSourceAsTree() {
+
     @TaskAction
     private void doInstall() {
+        getSourceAsTree().getElements();
```
Exactly. I should not create even draft PRs after working hours :)

donat commented on subprojects/gradle-guides-plugin/src/functionalTest/groovy/org/gradle/docs/samples/SamplesPluginFunctionalTest.groovy:
```diff
@@ -311,16 +311,37 @@ class SamplesPluginFunctionalTest extends AbstractSampleFunctionalSpec {
             |""".stripMargin()
     }
 
-    private void assertCanRunHelpTask(File zipFile) {
```
@big-guy This change should be part of #359. Rebasing...

donat replied:
✅

donat commented:
@big-guy I've addressed your concerns and added test coverage. I'm blocked by this PR so I'll do a merge in a minute. If you have further change requests I'll handle those separately. I hope it's fine like this.

donat commented:
That is if I could get a green build on CI.
